### Задание 1.
Написать теггер на данных с русским языком
проверить UnigramTagger, BigramTagger, TrigramTagger и их комбинации
написать свой теггер как на занятии, попробовать разные векторайзеры, добавить знание не только букв но и слов
сравнить все реализованные методы, сделать выводы

In [ ]:
!pip install pyconll

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install -U spacy
!python -m spacy info
!python -m spacy download ru_core_news_sm

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.7/6.7 MB 49.0 MB/s eta 0:00:00
  Attempting uninstall: spacy
    Found existing installation: spacy 3.4.4
    Uninstalling spacy-3.4.4:
      Successfully uninstalled spacy-3.4.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
en-core-web-sm 3.4.1 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.5.0 which is incompatible.
/usr/local/lib/python3.8/dist-packages/torch/cuda/__init__.py:497: UserWarning: Can't initialize NVML
  warnings.warn("Can't initialize NVML")
2023-01-25 17:44:02.784808: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
/usr/local/lib/python3.8/dist-packages/spacy/util.py:877: UserWarning: [W095] Model 'en_core_we

In [37]:
import pyconll

import pandas as pd

import nltk
from nltk.tag import UnigramTagger, BigramTagger, TrigramTagger

from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import CountVectorizer, HashingVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, accuracy_score, classification_report


In [ ]:
full_train = pyconll.load_from_file('ru_syntagrus-ud-train-a.conllu')
full_test = pyconll.load_from_file('ru_syntagrus-ud-dev.conllu')

In [ ]:
for sent in full_train[:2]:
    for token in sent:
        print(token.form, token.upos)
    print()

Анкета NOUN
. PUNCT

Начальник NOUN
областного ADJ
управления NOUN
связи NOUN
Семен PROPN
Еремеевич PROPN
был AUX
человек NOUN
простой ADJ
, PUNCT
приходил VERB
на ADP
работу NOUN
всегда ADV
вовремя ADV
, PUNCT
здоровался VERB
с ADP
секретаршей NOUN
за ADP
руку NOUN
и CCONJ
иногда ADV
даже PART
писал VERB
в ADP
стенгазету NOUN
заметки NOUN
под ADP
псевдонимом NOUN
" PUNCT
Муха NOUN
" PUNCT
. PUNCT



In [ ]:
fdata_train = []
for sent in full_train[:]:
    fdata_train.append([(token.form, token.upos) for token in sent])
    
fdata_test = []
for sent in full_test[:]:
    fdata_test.append([(token.form, token.upos) for token in sent])
    
fdata_sent_test = []
for sent in full_test[:]:
    fdata_sent_test.append([token.form for token in sent])

In [ ]:
MAX_SENT_LEN = max(len(sent) for sent in full_train)
MAX_ORIG_TOKEN_LEN = max(len(token.form) for sent in full_train for token in sent)
print('Наибольшая длина предложения', MAX_SENT_LEN)
print('Наибольшая длина токена', MAX_ORIG_TOKEN_LEN)

Наибольшая длина предложения 194
Наибольшая длина токена 31


In [ ]:
fdata_train[:3]

[[('Анкета', 'NOUN'), ('.', 'PUNCT')],
 [('Начальник', 'NOUN'),
  ('областного', 'ADJ'),
  ('управления', 'NOUN'),
  ('связи', 'NOUN'),
  ('Семен', 'PROPN'),
  ('Еремеевич', 'PROPN'),
  ('был', 'AUX'),
  ('человек', 'NOUN'),
  ('простой', 'ADJ'),
  (',', 'PUNCT'),
  ('приходил', 'VERB'),
  ('на', 'ADP'),
  ('работу', 'NOUN'),
  ('всегда', 'ADV'),
  ('вовремя', 'ADV'),
  (',', 'PUNCT'),
  ('здоровался', 'VERB'),
  ('с', 'ADP'),
  ('секретаршей', 'NOUN'),
  ('за', 'ADP'),
  ('руку', 'NOUN'),
  ('и', 'CCONJ'),
  ('иногда', 'ADV'),
  ('даже', 'PART'),
  ('писал', 'VERB'),
  ('в', 'ADP'),
  ('стенгазету', 'NOUN'),
  ('заметки', 'NOUN'),
  ('под', 'ADP'),
  ('псевдонимом', 'NOUN'),
  ('"', 'PUNCT'),
  ('Муха', 'NOUN'),
  ('"', 'PUNCT'),
  ('.', 'PUNCT')],
 [('В', 'ADP'),
  ('приемной', 'NOUN'),
  ('его', 'PRON'),
  ('с', 'ADP'),
  ('утра', 'NOUN'),
  ('ожидали', 'VERB'),
  ('посетители', 'NOUN'),
  (',', 'PUNCT'),
  ('-', 'PUNCT'),
  ('кое-кто', 'PRON'),
  ('с', 'ADP'),
  ('важными', 'ADJ'),

In [ ]:
unigram_tagger = UnigramTagger(fdata_train)
acc_u = unigram_tagger.evaluate(fdata_test)
display(acc_u)

<ipython-input-12-c0c1d20b533d>:2: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  acc_u = unigram_tagger.evaluate(fdata_test)


0.823732013802982

In [ ]:
bigram_tagger = BigramTagger(fdata_train)
acc_b = bigram_tagger.evaluate(fdata_test)
display(acc_b)

<ipython-input-14-8fa1fd4878f1>:2: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  acc_b = bigram_tagger.evaluate(fdata_test)


0.6093886320724006

In [ ]:
trigram_tagger = TrigramTagger(fdata_train)
acc_t = trigram_tagger.evaluate(fdata_test)
display(acc_t)

<ipython-input-16-921b35dd23f0>:2: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  acc_t = trigram_tagger.evaluate(fdata_test)


0.1778631421316492

In [ ]:
combination_1 = [BigramTagger, TrigramTagger]
combination_2 = [TrigramTagger, UnigramTagger]
combination_3 = [UnigramTagger, BigramTagger]
combination_4 = [BigramTagger, TrigramTagger, UnigramTagger]

In [ ]:
def backoff_tagger(train_sents, tagger_classes, backoff=None):
    for cls in tagger_classes:
        backoff = cls(train_sents, backoff=backoff)
    return backoff

acc_bo = []

for combination in [combination_1, combination_2, combination_3, combination_4]:
    backoff = nltk.DefaultTagger('NN') 
    tag = backoff_tagger(fdata_train,  
                         combination,  
                         backoff = backoff) 

    acc_bo.append(tag.evaluate(fdata_test))

acc_bo

<ipython-input-22-a1b8d409cd35>:14: DeprecationWarning: 
  Function evaluate() has been deprecated.  Use accuracy(gold)
  instead.
  acc_bo.append(tag.evaluate(fdata_test))


[0.1750309264926102,
 0.21248779217396965,
 0.8275343446838986,
 0.3623152548994075]

In [ ]:
result = pd.DataFrame({'Tagger': ['UnigramTagger', 'BigramTagger', 'TrigramTagger', 'BigramTagger & TrigramTagger', 'TrigramTagger & UnigramTagger', 'UnigramTagger & BigramTagger', 'All taggers'], 'Accuracy' : [acc_u, acc_b, acc_t] + acc_bo})
result.sort_values('Accuracy', ascending=False)

Tagger  Accuracy
5   UnigramTagger & BigramTagger  0.827534
0                  UnigramTagger  0.823732
1                   BigramTagger  0.609389
6                    All taggers  0.362315
4  TrigramTagger & UnigramTagger  0.212488
2                  TrigramTagger  0.177863
3   BigramTagger & TrigramTagger  0.175031

In [ ]:
train_tok = []
train_label = []
for sent in fdata_train[:]:
    for tok in sent:
        train_tok.append(tok[0])
        train_label.append('NO_TAG' if tok[1] is None else tok[1])
        
test_tok = []
test_label = []
for sent in fdata_test[:]:
    for tok in sent:
        test_tok.append(tok[0])
        test_label.append('NO_TAG' if tok[1] is None else tok[1])

In [ ]:
le = LabelEncoder()
train_enc_labels = le.fit_transform(train_label)
test_enc_labels = le.transform(test_label)
le.classes_

array(['ADJ', 'ADP', 'ADV', 'AUX', 'CCONJ', 'DET', 'INTJ', 'NOUN',
       'NO_TAG', 'NUM', 'PART', 'PRON', 'PROPN', 'PUNCT', 'SCONJ', 'SYM',
       'VERB', 'X'], dtype='<U6')

In [ ]:
vectorizers = [CountVectorizer(ngram_range=(1, 3), analyzer='char'), 
               TfidfVectorizer(ngram_range=(1, 3), analyzer='char'), 
               HashingVectorizer(ngram_range=(1, 3), analyzer='char', n_features=1000)] 
vectorizers_word = [CountVectorizer(ngram_range=(1, 3), analyzer='word'), 
               TfidfVectorizer(ngram_range=(1, 3), analyzer='word'), 
               HashingVectorizer(ngram_range=(1, 3), analyzer='word', n_features=1000)] 
n_features = [2000, 3000, 5000, 10000]
hvectorizer = [HashingVectorizer(ngram_range=(1, 3), analyzer='char', n_features=feat) for feat in n_features]
hvectorizer_word = [HashingVectorizer(ngram_range=(1, 3), analyzer='word', n_features=feat) for feat in n_features]

In [ ]:
f1_scores = []
acc_scores = []

for vectorizer in vectorizers + vectorizers_word + hvectorizer + hvectorizer_word:
    X_train = vectorizer.fit_transform(train_tok)
    X_test = vectorizer.transform(test_tok[:115000])
    
    lr = LogisticRegression(random_state=0, max_iter=100)
    lr.fit(X_train, train_enc_labels)
    pred = lr.predict(X_test)
    f1 = f1_score(test_enc_labels[:115000], pred, average='weighted')
    f1_scores.append(f1)
    acc = accuracy_score(test_enc_labels[:115000], pred)
    acc_scores.append(acc)
    
    print(vectorizer)
    print(classification_report(test_enc_labels[:115000], pred, target_names=le.classes_))

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


CountVectorizer(analyzer='char', ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.92      0.91      0.92     11247
         ADP       0.98      1.00      0.99     10255
         ADV       0.92      0.90      0.91      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.98      0.93      4276
         DET       0.88      0.75      0.81      2978
        INTJ       0.33      0.36      0.35        11
        NOUN       0.92      0.95      0.94     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.86      0.90      0.88      1436
        PART       0.95      0.78      0.86      3762
        PRON       0.83      0.89      0.86      5346
       PROPN       0.79      0.59      0.67      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.91      0.86      2176
         SYM       1.00      0.68      0.81        53
        VERB       0.94     

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


TfidfVectorizer(analyzer='char', ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.90      0.91      0.91     11247
         ADP       0.99      0.99      0.99     10255
         ADV       0.92      0.86      0.89      5986
         AUX       0.81      0.97      0.89      1058
       CCONJ       0.88      0.98      0.93      4276
         DET       0.80      0.83      0.82      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.90      0.96      0.93     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.85      0.90      0.87      1436
        PART       0.95      0.79      0.86      3762
        PRON       0.87      0.84      0.86      5346
       PROPN       0.80      0.52      0.63      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.91      0.86      2176
         SYM       1.00      0.64      0.78        53
        VERB       0.93      0.93      0.93     12617
           X       0.45    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


HashingVectorizer(analyzer='char', n_features=1000, ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.84      0.84      0.84     11247
         ADP       0.97      0.98      0.98     10255
         ADV       0.83      0.79      0.81      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.97      0.93      4276
         DET       0.80      0.80      0.80      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.84      0.90      0.87     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.81      0.82      0.82      1436
        PART       0.92      0.78      0.84      3762
        PRON       0.84      0.86      0.85      5346
       PROPN       0.72      0.45      0.55      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.90      0.85      2176
         SYM       1.00      0.64      0.78        53
        VERB       0.88      0.84      0.86     12617
           X       0.31    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


CountVectorizer(ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.95      0.37      0.54     11247
         ADP       0.99      0.48      0.65     10255
         ADV       0.93      0.78      0.85      5986
         AUX       0.83      0.88      0.85      1058
       CCONJ       0.90      0.20      0.33      4276
         DET       0.83      0.69      0.75      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.98      0.68      0.80     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.89      0.54      0.67      1436
        PART       0.97      0.75      0.84      3762
        PRON       0.81      0.82      0.82      5346
       PROPN       0.94      0.14      0.24      4315
       PUNCT       0.37      1.00      0.54     21941
       SCONJ       0.78      0.78      0.78      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.97      0.43      0.60     12617
           X       0.00    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


TfidfVectorizer(ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.95      0.44      0.60     11247
         ADP       0.99      0.48      0.65     10255
         ADV       0.96      0.80      0.87      5986
         AUX       0.83      0.88      0.85      1058
       CCONJ       0.89      0.20      0.33      4276
         DET       0.75      0.84      0.79      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.98      0.68      0.80     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.90      0.52      0.66      1436
        PART       0.97      0.75      0.84      3762
        PRON       0.88      0.76      0.81      5346
       PROPN       0.94      0.14      0.24      4315
       PUNCT       0.38      1.00      0.55     21941
       SCONJ       0.79      0.85      0.82      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.97      0.49      0.65  

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(n_features=1000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.43      0.21      0.28     11247
         ADP       0.83      0.47      0.60     10255
         ADV       0.59      0.63      0.61      5986
         AUX       0.70      0.94      0.80      1058
       CCONJ       0.84      0.18      0.30      4276
         DET       0.49      0.55      0.52      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.25      0.53      0.34     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.41      0.43      0.42      1436
        PART       0.86      0.76      0.81      3762
        PRON       0.64      0.76      0.70      5346
       PROPN       0.32      0.08      0.13      4315
       PUNCT       0.00      0.00      0.00     21941
       SCONJ       0.67      0.95      0.78      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.45   

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(analyzer='char', n_features=2000, ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.87      0.87      0.87     11247
         ADP       0.98      0.99      0.98     10255
         ADV       0.87      0.82      0.84      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.97      0.93      4276
         DET       0.84      0.75      0.80      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.86      0.93      0.89     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.83      0.84      0.83      1436
        PART       0.94      0.78      0.85      3762
        PRON       0.83      0.89      0.85      5346
       PROPN       0.74      0.41      0.53      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.90      0.85      2176
         SYM       1.00      0.64      0.78        53
        VERB       0.89      0.88      0.89     12617
           X       0.38    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


HashingVectorizer(analyzer='char', n_features=3000, ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.87      0.87      0.87     11247
         ADP       0.98      0.99      0.98     10255
         ADV       0.88      0.82      0.85      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.98      0.93      4276
         DET       0.82      0.80      0.81      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.87      0.93      0.90     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.83      0.83      0.83      1436
        PART       0.95      0.77      0.85      3762
        PRON       0.84      0.87      0.85      5346
       PROPN       0.75      0.40      0.52      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.90      0.85      2176
         SYM       1.00      0.79      0.88        53
        VERB       0.88      0.89      0.89     12617
           X       0.28    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


HashingVectorizer(analyzer='char', n_features=5000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.88      0.88      0.88     11247
         ADP       0.98      0.99      0.98     10255
         ADV       0.90      0.83      0.86      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.97      0.93      4276
         DET       0.83      0.78      0.81      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.87      0.94      0.90     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.84      0.87      0.85      1436
        PART       0.94      0.78      0.85      3762
        PRON       0.83      0.86      0.85      5346
       PROPN       0.76      0.41      0.54      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.81      0.90      0.85      2176
         SYM       1.00      0.64      0.78        53
        V

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(analyzer='char', n_features=10000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.88      0.88      0.88     11247
         ADP       0.98      0.99      0.98     10255
         ADV       0.90      0.84      0.87      5986
         AUX       0.81      0.97      0.88      1058
       CCONJ       0.88      0.97      0.93      4276
         DET       0.83      0.78      0.81      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.88      0.94      0.91     27241
      NO_TAG       1.00      1.00      1.00       197
         NUM       0.84      0.88      0.86      1436
        PART       0.94      0.79      0.85      3762
        PRON       0.82      0.87      0.85      5346
       PROPN       0.78      0.42      0.54      4315
       PUNCT       1.00      1.00      1.00     21941
       SCONJ       0.82      0.86      0.84      2176
         SYM       1.00      0.64      0.78        53
        

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(n_features=2000, ngram_range=(1, 3))


/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

         ADJ       0.50      0.26      0.34     11247
         ADP       0.90      0.48      0.62     10255
         ADV       0.68      0.69      0.68      5986
         AUX       0.75      0.94      0.84      1058
       CCONJ       0.89      0.18      0.31      4276
         DET       0.67      0.53      0.59      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.61      0.58      0.59     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.53      0.47      0.50      1436
        PART       0.91      0.76      0.83      3762
        PRON       0.69      0.85      0.76      5346
       PROPN       0.39      0.10      0.15      4315
       PUNCT       0.48      1.00      0.65     21941
       SCONJ       0.76      0.90      0.82      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.55      0.31      0.39     12617
           X       0.00    

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


HashingVectorizer(n_features=3000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.55      0.30      0.39     11247
         ADP       0.92      0.48      0.63     10255
         ADV       0.77      0.71      0.74      5986
         AUX       0.77      0.94      0.85      1058
       CCONJ       0.93      0.18      0.30      4276
         DET       0.71      0.57      0.63      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.65      0.60      0.62     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.61      0.49      0.54      1436
        PART       0.90      0.78      0.84      3762
        PRON       0.74      0.83      0.79      5346
       PROPN       0.44      0.12      0.18      4315
       PUNCT       0.47      1.00      0.64     21941
       SCONJ       0.73      0.95      0.83      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.60   

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(n_features=5000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.63      0.33      0.44     11247
         ADP       0.91      0.48      0.63     10255
         ADV       0.82      0.72      0.77      5986
         AUX       0.79      0.95      0.86      1058
       CCONJ       0.92      0.19      0.31      4276
         DET       0.69      0.70      0.70      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.70      0.62      0.65     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.66      0.48      0.55      1436
        PART       0.91      0.76      0.83      3762
        PRON       0.79      0.79      0.79      5346
       PROPN       0.54      0.13      0.21      4315
       PUNCT       0.45      1.00      0.62     21941
       SCONJ       0.76      0.88      0.82      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.67   

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative sol

HashingVectorizer(n_features=10000, ngram_range=(1, 3))
              precision    recall  f1-score   support

         ADJ       0.72      0.37      0.48     11247
         ADP       0.96      0.48      0.64     10255
         ADV       0.88      0.75      0.81      5986
         AUX       0.81      0.88      0.84      1058
       CCONJ       0.88      0.20      0.33      4276
         DET       0.72      0.75      0.74      2978
        INTJ       0.00      0.00      0.00        11
        NOUN       0.76      0.64      0.70     27241
      NO_TAG       0.00      0.00      0.00       197
         NUM       0.73      0.57      0.64      1436
        PART       0.96      0.75      0.84      3762
        PRON       0.83      0.78      0.81      5346
       PROPN       0.67      0.15      0.24      4315
       PUNCT       0.43      1.00      0.60     21941
       SCONJ       0.78      0.88      0.83      2176
         SYM       0.00      0.00      0.00        53
        VERB       0.75  

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
result_model = pd.DataFrame({'Vectorizer': vectorizers + vectorizers_word + hvectorizer + hvectorizer_word,
                            'f1_score': f1_scores})
result_model.sort_values('f1_score', ascending=False)

Vectorizer  f1_score
0   CountVectorizer(analyzer='char', ngram_range=(...  0.927826
1   TfidfVectorizer(analyzer='char', ngram_range=(...  0.921158
9   HashingVectorizer(analyzer='char', n_features=...  0.903641
8   HashingVectorizer(analyzer='char', n_features=...  0.901192
7   HashingVectorizer(analyzer='char', n_features=...  0.896841
6   HashingVectorizer(analyzer='char', n_features=...  0.895273
2   HashingVectorizer(analyzer='char', n_features=...  0.882224
4                 TfidfVectorizer(ngram_range=(1, 3))  0.666235
3                 CountVectorizer(ngram_range=(1, 3))  0.649712
13  HashingVectorizer(n_features=10000, ngram_rang...  0.620867
12  HashingVectorizer(n_features=5000, ngram_range...  0.594828
11  HashingVectorizer(n_features=3000, ngram_range...  0.577216
10  HashingVectorizer(n_features=2000, ngram_range...  0.555762
5   HashingVectorizer(n_features=1000, ngram_range...  0.345039

In [ ]:
result_model_acc = pd.DataFrame({'Vectorizer': vectorizers + vectorizers_word + hvectorizer + hvectorizer_word,
                            'Accuracy': acc_scores})
result_model_acc.sort_values('Accuracy', ascending=False)

Vectorizer  Accuracy
0   CountVectorizer(analyzer='char', ngram_range=(...  0.929504
1   TfidfVectorizer(analyzer='char', ngram_range=(...  0.923583
9   HashingVectorizer(analyzer='char', n_features=...  0.907417
8   HashingVectorizer(analyzer='char', n_features=...  0.904974
7   HashingVectorizer(analyzer='char', n_features=...  0.900687
6   HashingVectorizer(analyzer='char', n_features=...  0.898939
2   HashingVectorizer(analyzer='char', n_features=...  0.885165
4                 TfidfVectorizer(ngram_range=(1, 3))  0.656504
3                 CountVectorizer(ngram_range=(1, 3))  0.641139
13  HashingVectorizer(n_features=10000, ngram_rang...  0.630722
12  HashingVectorizer(n_features=5000, ngram_range...  0.612913
11  HashingVectorizer(n_features=3000, ngram_range...  0.601217
10  HashingVectorizer(n_features=2000, ngram_range...  0.584009
5   HashingVectorizer(n_features=1000, ngram_range...  0.365243

Судя по результатам комбинации таггеров лидирует по Accuracy смесь из UnigramTagger & BigramTagger с результатом - 0.93
Во втором и третьем тестах лидирует CountVectorizer с f-score 0.927826 и accuracy 0.929504

### Задание 2. 
Проверить, насколько хорошо работает NER
Данные брать из http://www.labinform.ru/pub/named_entities/
проверить NER из nltk/spacy/deeppavlov.
написать свой NER, попробовать разные подходы.
передаём в сетку токен и его соседей.
передаём в сетку только токен.
свой вариант.
сравнить свои реализованные подходы на качество — вывести precision/recall/f1_score.


In [50]:
!pip install -U spacy
!python -m spacy info
!python -m spacy download ru_core_news_sm

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.7/6.7 MB 55.1 MB/s eta 0:00:00
  Attempting uninstall: spacy
    Found existing installation: spacy 3.4.4
    Uninstalling spacy-3.4.4:
      Successfully uninstalled spacy-3.4.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
en-core-web-sm 3.4.1 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.5.0 which is incompatible.
/usr/local/lib/python3.8/dist-packages/spacy/util.py:877: UserWarning: [W095] Model 'en_core_web_sm' (3.4.1) was trained with spaCy v3.4 and may not be 100% compatible with the current version (3.5.0). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy

In [59]:
!pip install navec

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [67]:
!pip install slovnet

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 KB 2.0 MB/s eta 0:00:00


In [70]:
!pip install ipymarkup

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26118 sha256=5daed93618eb5790cdeebb836da44a2c5161086ab3ae7f118fb5d39ef5ef778a
  Stored in directory: /root/.cache/pip/wheels/45/23/de/5789a92962483fd33cb06674792b9697c1b3766d7c7742830e
Successfully built intervaltree
  Attempting uninstall: intervaltree
    Found existing installation: intervaltree 2.1.0
    Uninstalling intervaltree-2.1.0:
      Successfully uninstalled intervaltree-2.1.0


In [111]:
!pip install sklearn_crfsuite

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 17.7 MB/s eta 0:00:00


In [83]:
!pip install patool

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.5/77.5 KB 3.9 MB/s eta 0:00:00


In [112]:
import os
import glob

import csv
from razdel import tokenize

from spacy import displacy
import ru_core_news_sm

from navec import Navec
from slovnet import NER
from ipymarkup import show_span_ascii_markup as show_markup

import sklearn_crfsuite
from sklearn_crfsuite import metrics

In [88]:
import patoolib
patoolib.extract_archive("collection3.zip", outdir="collection3")

patool: Extracting collection3.zip ...
patool: running /usr/bin/7z x -ocollection3 -- collection3.zip
patool: ... collection3.zip extracted to `collection3'.


'collection3'

In [41]:
fileDir = r"collection3"
fileExt = r".txt"
documents_txt = [_ for _ in os.listdir(fileDir) if _.endswith(fileExt)]
text_list = []
for file in documents_txt:
    doc = open('collection3/' + file, encoding='utf-8')
    text = doc.read()
    text_list.append(text)
    
data_text = pd.DataFrame({'text': text_list })
data_text

text
0    Совмин Белоруссии надеется разгрузить склады в...
1    Продолжаются увольнения энергетиков после крит...
2    \nЭкс-Президенту Египта Мурси продлили срок ар...
3    Медведев уволил Суходольского с должности глав...
4    Сын А.Лукашенко взял нити силовых структур в с...
..                                                 ...
719  СМИ назвали имя лидера "Октябрьского джамаата"...
720  Бывший председатель Банка России сменит Кудрин...
721  Антонова приветствует назначение Лошак на пост...
722  У пензенского Роскомнадзора новый глава\n\t\nА...
723  В.Путин уволил замминистра промышленности и то...

[724 rows x 1 columns]

In [42]:
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [43]:
document = data_text.text[0]
nltk.pos_tag(nltk.word_tokenize(document))

[('Совмин', 'JJ'),
 ('Белоруссии', 'NNP'),
 ('надеется', 'NNP'),
 ('разгрузить', 'NNP'),
 ('склады', 'NNP'),
 ('в', 'NNP'),
 ('Сибири', 'NNP'),
 ('Скопившаяся', 'NNP'),
 ('на', 'NNP'),
 ('складах', 'NNP'),
 ('белорусских', 'NNP'),
 ('предприятий', 'NNP'),
 ('продукция', 'NNP'),
 ('может', 'NNP'),
 ('быть', 'NNP'),
 ('реализована', 'NNP'),
 ('в', 'NNP'),
 ('Сибири', 'NNP'),
 ('при', 'NNP'),
 ('помощи', 'NNP'),
 ('административного', 'NNP'),
 ('ресурса', 'NNP'),
 ('российских', 'NNP'),
 ('чиновников', 'NNP'),
 ('.', '.'),
 ('Об', 'VB'),
 ('этом', 'JJ'),
 ('17', 'CD'),
 ('июня', 'JJ'),
 ('заявил', 'NNP'),
 ('председатель', 'NNP'),
 ('правительства', 'NNP'),
 ('Белоруссии', 'NNP'),
 ('Михаил', 'NNP'),
 ('Мясникович', 'NNP'),
 ('на', 'NNP'),
 ('встрече', 'NNP'),
 ('с', 'NNP'),
 ('делегацией', 'NNP'),
 ('Сибирского', 'NNP'),
 ('федерального', 'NNP'),
 ('округа', 'NNP'),
 ('России', 'NNP'),
 ('во', 'NNP'),
 ('главе', 'NNP'),
 ('с', 'NNP'),
 ('полномочным', 'NNP'),
 ('представителем', 'NNP'),


In [44]:
nltk.download('maxent_ne_chunker')
nltk.download('words')

[nltk_data] Downloading package maxent_ne_chunker to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping chunkers/maxent_ne_chunker.zip.
[nltk_data] Downloading package words to /root/nltk_data...
[nltk_data]   Unzipping corpora/words.zip.


True

In [45]:
{(' '.join(c[0] for c in chunk), chunk.label() ) for chunk in nltk.ne_chunk(nltk.pos_tag(nltk.word_tokenize(document))) if hasattr(chunk, 'label') }

{('Белоруссии', 'ORGANIZATION'),
 ('Белоруссии', 'PERSON'),
 ('Лукашенко', 'PERSON'),
 ('Минске Александр Лукашенко', 'PERSON'),
 ('Михаил Мясникович', 'PERSON'),
 ('Напомним', 'PERSON'),
 ('России', 'PERSON'),
 ('Сибири', 'PERSON'),
 ('Сибири Скопившаяся', 'PERSON'),
 ('Совмин', 'PERSON'),
 ('Совмина Белоруссии', 'PERSON')}

In [47]:
pd.read_csv('collection3/001.ann', delimiter='\t')

T1      LOC 0 6            Россия
0    T2    LOC 50 53               США
1    T3    LOC 57 63            Грузию
2    T4    LOC 87 93            МОСКВА
3    T5  ORG 103 114       РИА Новости
4    T6  LOC 116 122            Россия
5    T7  LOC 141 144               США
6    T8  LOC 161 168           Тбилиси
7    T9  LOC 301 307            России
8   T10  PER 308 324  Григорий Карасин
9   T11  LOC 383 386               США
10  T12  PER 387 402   Дэниэлом Фридом
11  T13  LOC 505 517      Южной Осетии
12  T14  LOC 703 709            Россия
13  T15  LOC 723 730           Тбилиси
14  T16  LOC 815 825        Вашингтона
15  T17  ORG 838 841               МИД
16  T18  LOC 842 848            России

In [54]:
nlp = ru_core_news_sm.load()
ny_bb = data_text.text[0]
article = nlp(ny_bb)
displacy.render(article, jupyter=True, style='ent')

In [55]:
for token in article:
    print(token.text, token.pos_, token.dep_)

Совмин NOUN nsubj
Белоруссии PROPN nmod
надеется VERB ROOT
разгрузить VERB xcomp
склады NOUN obj
в ADP case
Сибири PROPN nmod


 SPACE dep
Скопившаяся VERB acl
на ADP case
складах NOUN obl
белорусских ADJ amod
предприятий NOUN nmod
продукция NOUN nsubj:pass
может VERB parataxis
быть AUX aux:pass
реализована VERB xcomp
в ADP case
Сибири PROPN obl
при ADP case
помощи NOUN fixed
административного ADJ amod
ресурса NOUN obl
российских ADJ amod
чиновников NOUN nmod
. PUNCT punct
Об ADP case
этом PRON obl
17 ADJ obl
июня NOUN flat
заявил VERB ROOT
председатель NOUN nsubj
правительства NOUN nmod
Белоруссии PROPN nmod
Михаил PROPN appos
Мясникович PROPN flat:name
на ADP case
встрече NOUN obl
с ADP case
делегацией NOUN nmod
Сибирского ADJ amod
федерального ADJ amod
округа NOUN nmod
России PROPN nmod
во ADP case
главе NOUN fixed
с ADP case
полномочным ADJ amod
представителем NOUN obl
президента NOUN nmod
РФ PROPN nmod
Виктором PROPN appos
Толоконским PROPN flat:name
. PUNCT punct


 SPACE dep
Во 

In [56]:
pd.read_csv('collection3/001.ann', delimiter='\t', header=None)

0            1                 2
0    T1      LOC 0 6            Россия
1    T2    LOC 50 53               США
2    T3    LOC 57 63            Грузию
3    T4    LOC 87 93            МОСКВА
4    T5  ORG 103 114       РИА Новости
5    T6  LOC 116 122            Россия
6    T7  LOC 141 144               США
7    T8  LOC 161 168           Тбилиси
8    T9  LOC 301 307            России
9   T10  PER 308 324  Григорий Карасин
10  T11  LOC 383 386               США
11  T12  PER 387 402   Дэниэлом Фридом
12  T13  LOC 505 517      Южной Осетии
13  T14  LOC 703 709            Россия
14  T15  LOC 723 730           Тбилиси
15  T16  LOC 815 825        Вашингтона
16  T17  ORG 838 841               МИД
17  T18  LOC 842 848            России

In [72]:
text = data_text.text[0]
navec = Navec.load('navec_news_v1_1B_250K_300d_100q.tar')
ner = NER.load('slovnet_ner_news_v1.tar')
ner.navec(navec)

markup = ner(text)
show_markup(markup.text, markup.spans)

Совмин Белоруссии надеется разгрузить склады в Сибири
ORG─── LOC───────                              LOC───
Скопившаяся на складах белорусских предприятий продукция может быть 
реализована в Сибири при помощи административного ресурса российских 
              LOC───                                                 
чиновников. Об этом 17 июня заявил председатель правительства 
Белоруссии Михаил Мясникович на встрече с делегацией Сибирского 
LOC─────── PER──────────────                         LOC────────
федерального округа России во главе с полномочным представителем 
─────────────────── LOC───                                       
президента РФ Виктором Толоконским.
           LO PER───────────────── 
Во время встречи Мясникович указал на "огромный потенциал" 
                 PER───────                                
сотрудничества Белоруссии с Сибирским федеральным округом: по итогам 
               LOC───────   LOC──────────────────────────            
2012 года взаимный товароо

In [73]:
fileDir = r"collection3"
fileExt = r".ann"
documents_ann = [_ for _ in os.listdir(fileDir) if _.endswith(fileExt)]
print(documents_ann)

['356.ann', '149.ann', '368.ann', '183.ann', '252.ann', '094.ann', '100.ann', '537.ann', '409.ann', '130.ann', '623.ann', 'last_21.ann', '428.ann', '251.ann', '434.ann', '375.ann', '483.ann', '328.ann', '171.ann', '357.ann', 'artjakov.ann', '091.ann', '445.ann', '514.ann', '14_01_13i.ann', '279.ann', '058.ann', '374.ann', '299.ann', '042.ann', '243.ann', '263.ann', '431.ann', '310.ann', '362.ann', '25_12_12a.ann', '032.ann', '316.ann', '399.ann', '288.ann', '324.ann', '063.ann', '493.ann', '270.ann', '439.ann', '22_11_12d.ann', '192.ann', '170.ann', '118.ann', '446.ann', '506.ann', '309.ann', '23_11_12b.ann', '191.ann', '072.ann', '045.ann', '449.ann', '575.ann', '109.ann', '093.ann', '21_11_12c.ann', '20_11_12i.ann', '244.ann', '410.ann', '412.ann', '326.ann', '461.ann', '014.ann', '343.ann', '10_01_13a.ann', '146.ann', '377.ann', '038.ann', '09_01_13c.ann', '532.ann', '281.ann', '119.ann', '061.ann', '067.ann', '029.ann', '459.ann', 'chaves.ann', '286.ann', '289.ann', '240.ann', '233

In [74]:
ann = pd.read_csv('collection3/001.ann', delimiter='\t', header=None)
ann

0            1                 2
0    T1      LOC 0 6            Россия
1    T2    LOC 50 53               США
2    T3    LOC 57 63            Грузию
3    T4    LOC 87 93            МОСКВА
4    T5  ORG 103 114       РИА Новости
5    T6  LOC 116 122            Россия
6    T7  LOC 141 144               США
7    T8  LOC 161 168           Тбилиси
8    T9  LOC 301 307            России
9   T10  PER 308 324  Григорий Карасин
10  T11  LOC 383 386               США
11  T12  PER 387 402   Дэниэлом Фридом
12  T13  LOC 505 517      Южной Осетии
13  T14  LOC 703 709            Россия
14  T15  LOC 723 730           Тбилиси
15  T16  LOC 815 825        Вашингтона
16  T17  ORG 838 841               МИД
17  T18  LOC 842 848            России

In [98]:
docs = []
# print(len(data_text))
# print(len(documents_ann))
# for i in range(len(data_text)):
for i in range(len(documents_ann)):
    words = []
    labels = []
    text = data_text['text'][i]
    # print(documents_ann[i])
    df = pd.read_csv('Collection3/' + documents_ann[i], delimiter='\t', header=None, quoting=csv.QUOTE_NONE, encoding='utf-8')
    df_ann = pd.DataFrame()
    df_ann['Token'] = df.loc[:, 2]
    split_1 = [loc.split() for loc in df.loc[:, 1].values]
    df_ann['Entity'] = [loc[0] for loc in split_1]
       
    dic = {}
    for j in range(len(df)):
        token = df_ann['Token'][j].lower().split()
        entity = df_ann['Entity'][j]
        for tok in token:
            dic[tok] = entity

    for token in tokenize(text):
        if (token.text.lower() in dic.keys()):
            words.append(token.text)
            labels.append(dic[token.text.lower()])
        else:
            words.append(token.text)
            labels.append('OUT')
    
    docs.append([words, labels])

In [100]:
print(docs[0][0]), print(docs[0][1])

['Совмин', 'Белоруссии', 'надеется', 'разгрузить', 'склады', 'в', 'Сибири', 'Скопившаяся', 'на', 'складах', 'белорусских', 'предприятий', 'продукция', 'может', 'быть', 'реализована', 'в', 'Сибири', 'при', 'помощи', 'административного', 'ресурса', 'российских', 'чиновников', '.', 'Об', 'этом', '17', 'июня', 'заявил', 'председатель', 'правительства', 'Белоруссии', 'Михаил', 'Мясникович', 'на', 'встрече', 'с', 'делегацией', 'Сибирского', 'федерального', 'округа', 'России', 'во', 'главе', 'с', 'полномочным', 'представителем', 'президента', 'РФ', 'Виктором', 'Толоконским', '.', 'Во', 'время', 'встречи', 'Мясникович', 'указал', 'на', '"', 'огромный', 'потенциал', '"', 'сотрудничества', 'Белоруссии', 'с', 'Сибирским', 'федеральным', 'округом', ':', 'по', 'итогам', '2012', 'года', 'взаимный', 'товарооборот', 'составил', 'более', '$', '1,4', 'млрд', '-', 'на', '4,4', '%', 'больше', 'уровня', '2011', 'года', '.', 'Председатель', 'Совмина', 'Белоруссии', 'с', 'сожалением', 'отметил', ':', 'в', 'I

(None, None)

In [101]:
data, labels = list(zip(*docs))
for w, e in zip(data[0], labels[0]):
    print(f'{w}\t{e}')

Совмин	OUT
Белоруссии	OUT
надеется	OUT
разгрузить	OUT
склады	OUT
в	OUT
Сибири	OUT
Скопившаяся	OUT
на	OUT
складах	OUT
белорусских	OUT
предприятий	OUT
продукция	OUT
может	OUT
быть	OUT
реализована	OUT
в	OUT
Сибири	OUT
при	OUT
помощи	OUT
административного	OUT
ресурса	OUT
российских	OUT
чиновников	OUT
.	OUT
Об	OUT
этом	OUT
17	OUT
июня	OUT
заявил	OUT
председатель	OUT
правительства	OUT
Белоруссии	OUT
Михаил	OUT
Мясникович	OUT
на	OUT
встрече	OUT
с	OUT
делегацией	OUT
Сибирского	OUT
федерального	OUT
округа	OUT
России	OUT
во	OUT
главе	OUT
с	OUT
полномочным	OUT
представителем	OUT
президента	OUT
РФ	OUT
Виктором	OUT
Толоконским	OUT
.	OUT
Во	OUT
время	OUT
встречи	OUT
Мясникович	OUT
указал	OUT
на	OUT
"	OUT
огромный	OUT
потенциал	OUT
"	OUT
сотрудничества	OUT
Белоруссии	OUT
с	OUT
Сибирским	OUT
федеральным	OUT
округом	OUT
:	OUT
по	OUT
итогам	OUT
2012	OUT
года	OUT
взаимный	OUT
товарооборот	OUT
составил	OUT
более	OUT
$	OUT
1,4	OUT
млрд	OUT
-	OUT
на	OUT
4,4	OUT
%	OUT
больше	OUT
уровня	OUT
2011	OUT
года	OUT


In [102]:
df = pd.DataFrame({'sent_id': [i for j in [[i] * len(s) for i, s in enumerate(data)] for i in j],
                   'data': [i for j in data for i in j],
                   'entities': [i for j in labels for i in j]})
df.head(50)

sent_id               data entities
0         0             Совмин      OUT
1         0         Белоруссии      OUT
2         0           надеется      OUT
3         0         разгрузить      OUT
4         0             склады      OUT
5         0                  в      OUT
6         0             Сибири      OUT
7         0        Скопившаяся      OUT
8         0                 на      OUT
9         0            складах      OUT
10        0        белорусских      OUT
11        0        предприятий      OUT
12        0          продукция      OUT
13        0              может      OUT
14        0               быть      OUT
15        0        реализована      OUT
16        0                  в      OUT
17        0             Сибири      OUT
18        0                при      OUT
19        0             помощи      OUT
20        0  административного      OUT
21        0            ресурса      OUT
22        0         российских      OUT
23        0         чиновников      OUT
24        0                  .      OUT
25        0                 Об      OUT
26        0               этом      OUT
27        0                 17      OUT
28        0               июня      OUT
29        0             заявил      OUT
30        0       председатель      OUT
31        0      правительства      OUT
32        0         Белоруссии      OUT
33        0             Михаил      OUT
34        0         Мясникович      OUT
35        0                 на      OUT
36        0            встрече      OUT
37        0                  с      OUT
38        0         делегацией      OUT
39        0         Сибирского      OUT
40        0       федерального      OUT
41        0             округа      OUT
42        0             России      OUT
43        0                 во      OUT
44        0              главе      OUT
45        0                  с      OUT
46        0        полномочным      OUT
47        0     представителем      OUT
48        0         президента      OUT
49        0                 РФ      OUT

In [103]:
class SentenceGetter(object):
    
    def __init__(self, data):
        self.n_sent = 1
        self.data = data
        self.empty = False
        agg_func = lambda s: [(w, t) for w, t in zip(s['data'].values.tolist(), 
                                                           s['entities'].values.tolist())]
        self.grouped = self.data.groupby('sent_id').apply(agg_func)
        self.sentences = [s for s in self.grouped]
        
    def get_next(self):
        try: 
            s = self.grouped['Sentence: {}'.format(self.n_sent)]
            self.n_sent += 1
            return s 
        except:
            return None

getter = SentenceGetter(df)

sentences = getter.sentences

In [104]:
def word2features(sent, i):
    word = sent[i][0]
    postag = sent[i][1]
    
    features = {
        'bias': 1.0, 
        'word.lower()': word.lower(), 
        'word[-3:]': word[-3:],
        'word[-2:]': word[-2:],
        'word.isdigit()': word.isdigit()
    }
    if i > 0:
        word1 = sent[i - 1][0]
        features.update({
            '-1:word.lower()': word1.lower()
        })
    else:
        features['BOS'] = True
    if i < len(sent) - 1:
        word1 = sent[i + 1][0]
        features.update({
            '+1:word.lower()': word1.lower()
        })
    else:
        features['EOS'] = True

    return features

def sent2features(sent):
    return [word2features(sent, i) for i in range(len(sent))]

def sent2labels(sent):
    return [label for token, label in sent]

def sent2tokens(sent):
    return [token for token, label in sent]

In [105]:
X = [sent2features(s) for s in sentences]
y = [sent2labels(s) for s in sentences]

In [106]:
X = [sent2features(s) for s in sentences]
y = [sent2labels(s) for s in sentences]
X[2][7]

{'bias': 1.0,
 'word.lower()': 'на',
 'word[-3:]': 'на',
 'word[-2:]': 'на',
 'word.isdigit()': False,
 '-1:word.lower()': 'еще',
 '+1:word.lower()': '15'}

In [107]:
len(X)

713

In [108]:
X_train = X[:700]
X_test = X[700:]
y_train = y[:700]
y_test = y[700:]

In [115]:
%%time

crf = sklearn_crfsuite.CRF(
    algorithm='lbfgs', # Градиентный спуск с использованием метода L-BFGS
    c1=0.1, # Коэффициент для регуляризации L1
    c2=0.1, # Коэффициент для регуляризации L2
    max_iterations=1000, # Максимальное количество итераций
    all_possible_transitions=True, # Генерация объектов (не встречающихся в обучающих данных)
    verbose=False # Выключение режима тренировки
)

try:
    crf.fit(X_train, y_train)
except AttributeError:
    pass

CPU times: user 28 s, sys: 89 ms, total: 28.1 s
Wall time: 29.3 s


In [116]:
all_entities = sorted(df.entities.unique().tolist())
len(all_entities)

4

In [117]:
y_pred = crf.predict(X_test)
metrics.flat_f1_score(y_test, y_pred, average='weighted', labels=all_entities)

0.9853420963702066

In [118]:
def word2features(sent, i):
    word = sent[i][0]
    postag = sent[i][1]
    
    features = {
        'bias': 1.0, 
        'word.lower()': word.lower(), 
        'word[-3:]': word[-3:],
        'word[-2:]': word[-2:],
        'word.isdigit()': word.isdigit()
    }
    return features

def sent2features(sent):
    return [word2features(sent, i) for i in range(len(sent))]

def sent2labels(sent):
    return [label for token, label in sent]

def sent2tokens(sent):
    return [token for token, label in sent]

In [119]:
X = [sent2features(s) for s in sentences]
y = [sent2labels(s) for s in sentences]
X[1][7]

{'bias': 1.0,
 'word.lower()': 'путина',
 'word[-3:]': 'ина',
 'word[-2:]': 'на',
 'word.isdigit()': False}

In [120]:
len(X)

713

In [121]:
X_train = X[:700]
X_test = X[700:]
y_train = y[:700]
y_test = y[700:]

In [123]:
%%time

crf = sklearn_crfsuite.CRF(
    algorithm='lbfgs',
    c1=0.1,
    c2=0.1,
    max_iterations=1000,
    all_possible_transitions=True,
    verbose=False
)

try:
    crf.fit(X_train, y_train)
except AttributeError:
    pass

CPU times: user 32.9 s, sys: 101 ms, total: 33 s
Wall time: 33.3 s


In [124]:
all_entities = sorted(df.entities.unique().tolist())
len(all_entities)

4

In [125]:
y_pred = crf.predict(X_test)
metrics.flat_f1_score(y_test, y_pred, average='weighted', labels=all_entities)

0.9855229392577891